homeassistant/components/qbus/__init__.py

```diff
@@ -0,0 +1,61 @@
+"""The Qbus integration."""
+
+import logging
+
+from qbusmqttapi.factory import QbusMqttTopicFactory
+
+from homeassistant.components.mqtt import async_wait_for_mqtt_client, client as mqtt
+from homeassistant.core import HomeAssistant
+
+from .const import PLATFORMS
+from .coordinator import QbusConfigEntry, QbusDataCoordinator, QbusRuntimeData
+from .qbus import QbusConfigContainer
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: QbusConfigEntry) -> bool:
+    """Set up Qbus from a config entry."""
+    _LOGGER.debug("Loading entry %s", entry.entry_id)
+
+    if not await async_wait_for_mqtt_client(hass):
+        _LOGGER.error("MQTT integration not available")
+        return False
+
+    coordinator = QbusDataCoordinator(hass, entry)
+    entry.runtime_data = QbusRuntimeData(coordinator)
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Subscribe to Qbus config topic
+    config_topic = QbusMqttTopicFactory().get_config_topic()
+    _LOGGER.debug("Subscribing to %s", config_topic)
+    entry.async_on_unload(
+        await mqtt.async_subscribe(hass, config_topic, coordinator.config_received)
+    )
+
+    # Request Qbus config
+    config = await QbusConfigContainer.async_get_or_request_config(hass)
+
+    if config:
+        await coordinator.async_update_config(config)
```

Comment:
There is an early exit in `async_get_or_request_config` when the config already has been received. This can happen when you have multiple devices. The first one will publish to the topic, and will then store it in memory. Subsequent requests to  `async_get_or_request_config` will receive the config from memory, so the `coordinator.config_received` won't be triggered then.

We still need `coordinator.config_received` on its own, because the Qbus management software can still cause config changes to be published to that topic.

Reply:
But if the config was already received, `config_received` would have been called already too (which internally calls `async_update_config`), so why is `async_update_config` needed to be called here?

Reply:
Because the configuration received contains the configuration for all devices, but the coordinator only handles one device. So `async_update_config` extracts the device specicif configuration and does some device specific things. Maybe the method should be renamed to `async_refresh_device` or something?

Reply:
Done